# Do not use Array.forEach (no-array-foreach)

Prefer other functional constructs like map(), reduce(), for...of instead of Array.forEach

## Rule Details

Examples of **incorrect** code for this rule:

```js
const arr = [1, 2, 3];
const out = [];
arr.forEach(i => out.push(i + 1));
```

Examples of **correct** code for this rule:

```js
const arr = [1, 2, 3];
const out = arr.map(i => i + 1);
```

## When Not To Use It

Let's fill it if we encounter a situation where forEach is preferable.

## Further Reading

https://medium.com/front-end-weekly/stop-array-foreach-and-start-using-filter-map-some-reduce-functions-298b4dabfa09
