# Prefer async/await to Promise (no-promise)

Prefer async/await to Promise.then() or Promise.catch().


## Rule Details

Examples of **incorrect** code for this rule:

```js
function getData(url) {
  request(url)
    .then(data => {
      return data;
    })
    .catch(error => {
      return error;
    })
}
```

Examples of **correct** code for this rule:

```js
async function getData(url) {
  let output;
  try {
    output = await request(url);
  } catch (error) {
    output = error;
  }
  return output;
}
```

## When Not To Use It

Async/await can only be used in the context of an async function.

## Further Reading

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function
